File: src/main/java/me/oddlyoko/odoo/OdooUtil.java
package me.oddlyoko.odoo;

import com.intellij.openapi.util.Key;
import com.intellij.psi.PsiElement;
import com.intellij.psi.util.CachedValue;
import com.intellij.psi.util.CachedValueProvider;
import com.intellij.psi.util.CachedValuesManager;
import com.intellij.psi.util.PsiModificationTracker;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Supplier;

public final class OdooUtil {
    private OdooUtil() {}

    /**
     * Retrieve saved data from the cache
     *
     * @param psiElement the element to retrieve the data from
     * @param supplier   the supplier that will generate the data if it doesn't exist in the cache
     * @param depends    the elements that will update the cache if they change
     * @return the data
     * @param <E> the type of the data
     */
    public static <E> E getData(PsiElement psiElement, Supplier<E> supplier, Object... depends) {
        return getData(psiElement, null, supplier, depends);
    }

    /**
     * Retrieve saved data from the cache
     *
     * @param psiElement the element to retrieve the data from
     * @param param      the key that identifies the data in the map
     * @param supplier   the supplier that will generate the data if it doesn't exist in the cache
     * @param depends    the elements that will update the cache if they change
     * @return the data
     * @param <E> the type of the data
     * @param <P> the type of the parameter
     */
    public static <E, P> E getData(PsiElement psiElement, P param, Supplier<E> supplier, Object... depends) {
        CachedValuesManager manager = CachedValuesManager.getManager(psiElement.getProject());
        return getData(psiElement, manager.getKeyForClass(depends.getClass()), param, supplier, depends);
    }

    /**
     * Retrieve saved data from the cache
     *
     * @param psiElement the element to retrieve the data from
     * @param key        the key that identifies the map that contains the data
     * @param param      the key that identifies the data in the map
     * @param supplier   the supplier that will generate the data if it doesn't exist in the cache
     * @param depends    the elements that will update the cache if they change
     * @return the data
     * @param <E> the type of the data
     * @param <P> the type of the parameter
     */
    public static <E, P> E getData(PsiElement psiElement, Key<CachedValue<Map<P, E>>> key, P param, Supplier<E> supplier, Object... depends) {
        Map<P, E> map = CachedValuesManager.getCachedValue(psiElement, key, () -> {
            Object[] objs = depends != null && depends.length > 0 ? depends : new Object[] { PsiModificationTracker.MODIFICATION_COUNT };
            return CachedValueProvider.Result.create(new HashMap<>(), objs);
        });
        E value = map.get(param);
        if (value == null) {
            value = supplier.get();
            map.put(param, value);
        }
        return value;
    }
}

File: src/main/kotlin/me/oddlyoko/odoo/OdooUtil.kt
package me.oddlyoko.odoo

import com.intellij.openapi.util.Key
import com.intellij.psi.PsiElement
import com.intellij.psi.util.CachedValue
import com.intellij.psi.util.CachedValueProvider
import com.intellij.psi.util.CachedValuesManager
import com.intellij.psi.util.PsiModificationTracker

object OdooUtil {

    fun <E, P> getData(psiElement: PsiElement, param: P, vararg depends: Any, supplier: () -> E?): E? =
        getData(psiElement, CachedValuesManager.getManager(psiElement.project).getKeyForClass(supplier.javaClass), param, depends, supplier=supplier)

    fun <E, P> getData(psiElement: PsiElement, key: Key<CachedValue<Map<P, E?>>>, param: P, vararg depends: Any, supplier: () -> E?): E? {
        val map: MutableMap<P, E?> = CachedValuesManager.getCachedValue(psiElement, key, CachedValueProvider {
            val objs = depends.takeIf { it.isNotEmpty() } ?: arrayOf(PsiModificationTracker.MODIFICATION_COUNT)
            return@CachedValueProvider CachedValueProvider.Result.create(LinkedHashMap(), objs)
        }) as MutableMap<P, E?>

        return map.getOrElse(param) {
            val value = supplier()
            map[param] = value
            return value
        }
    }
}
